### ISSUE TITLE
Question on Factor Management in Homework 5

### ISSUE THREAD

**Issue by yksugi** [STAT545-UBC/Discussion#186]
In the Drop Oceania section, the description of the homework says: "Use a figure that includes a legend to further explore the effects of filtering data and/or changing factor levels."

It's not entirely clear to me what we are supposed to do for this one.  Could someone explain?

**Comment by MagdaPrice** [comment]
Hi @yosukesugishita,

I think you are being asked to **show us** what the effects are of cleaning up a factor using a graph. Take a look at the difference between the graphs that Jenny made [here](https://stat545-ubc.github.io/block014_factors.html) when the factor country was reordered by slope (though there was no legend in these graphs). 

So here's a quick example from a subset of 5 countries **BEFORE** extra country levels were dropped
![image](https://cloud.githubusercontent.com/assets/13424175/10717892/550fd4a6-7b22-11e5-8ba7-aafa0e0ea8cc.png)

and **AFTER** country levels were dropped. Look at the much nicer legend!
![image](https://cloud.githubusercontent.com/assets/13424175/10717895/67e1b720-7b22-11e5-9080-466642825ab0.png)

**Comment by jennybc** [comment]
Yes what @MagdaPrice is saying :)

**Comment by zhamel** [comment]
Using the following codes, one for which I only `filter` the data, and one for which I also `drop` the unused factors, I get the same graph and the same legend. `Oceania` NEVER appears in the legend.. is it normal?

Filter:
`g <- gtbl %>% filter(continent != "Oceania") %>% group_by(continent,year) %>% summarize(mean = mean(lifeExp))
ggplot(g, aes(x = year, y = mean, color = continent))  +
  geom_point() +
  ylab("life expectancy") `

Filter+drop:
`g <- gtbl %>% filter(continent != "Oceania") %>% group_by(continent,year) %>% summarize(mean = mean(lifeExp)) %>% droplevels()
ggplot(g, aes(x = year, y = mean, color = continent))  +
  geom_point() +
  geom_line() +
  ylab("life expectancy")`

**Comment by zhamel** [comment]
Using the following codes, one for which I only `filter` the data, and one for which I also `drop` the unused factors, I get the same graph and the same legend. `Oceania` NEVER appears in the legend.. is it normal?

(1) When filtering...

```{r}
g <- gtbl %>% filter(continent != "Oceania") %>% group_by(continent,year) %>% 
      summarize(mean = mean(lifeExp))
ggplot(g, aes(x = year, y = mean, color = continent))  +
  geom_point() +
  ylab("life expectancy")
```

(2) When droping unused factors...

```{r}
g <- gtbl %>% filter(continent != "Oceania") %>% group_by(continent,year)  %>% 
     summarize(mean = mean(lifeExp)) %>% droplevels()
ggplot(g, aes(x = year, y = mean, color = continent))  +
  geom_point() +
  geom_line() +
  ylab("life expectancy")
```

**Comment by jennybc** [comment]
In addition to filtering the data, you are using `group_by() + summarize()` which happens to drop empty groups. When it probably should not (https://github.com/hadley/dplyr/issues/341).

The behavior I'm asking you to explore is ... does filtering the data alone change the factor levels? And the answer is "no". In general, you need to assume control of the factor levels. The fact that empty levels are being dropped with your code is a convenient coincidence that you shouldn't rely on.

**Comment by zhamel** [comment]
When I check the number of levels (after using `filter()` alone) I still have all of them; however; in my graph they didn't appear. 

So I just tried another graph where I only used the `filter()` function.

```{r}
ggplot(gtbl %>% filter(continent == "Oceania"), aes(x = year, y = lifeExp, color = country))  +
  geom_point() +
  ylab("life expectancy") 
```
But here, I still only get those two countries only in the legend, unless I add `scale_colour_discrete(drop = FALSE)` to get:

```{r}
ggplot(gtbl %>% filter(continent == "Oceania"), aes(x = year, y = lifeExp, color = country))  +
  geom_point() +
  ylab("life expectancy") +
  scale_colour_discrete(drop = TRUE)
```

I understand what we are trying to show, but I just can't fgure out how to show it in a graph. Only by using `levels()`...

**Comment by jennybc** [comment]
Hmmm... I see what you mean. Apparently `ggplot2`'s behavior changed quite a while ago such that the default is to drop unused levels in such a legend, unless explicitly told not to via `drop = FALSE` in e.g. `scale_color_discrete()`. So do that, to show that the levels are indeed still there, but note in your report the helpful default behavior of `ggplot2`. I think I first crafted this exercise back in the days when I used `lattice`.
